### PR TITLE
adding roots function for jax frontend

### LINKED ISSUE #23179: roots

### PR DESCRIPTION
**Pull request by dash96**
# PR Description

## Related Issue

Close #23179 

## Checklist

- [x] Did you add a function?
- [x] Did you add the tests?
- [x] Did you follow the steps we provided?

### CODE REVIEW

**Comment by dash96** [comment]
@Ookamice please help me understand this error. Because when I use some random inputs to this function to verify it with jax.roots() it provides the the same results,

fun_name = 'roots', args = ([Array([-1., -1., -1., -1., -1.], dtype=float32)],)
pos = 0, arg = [Array([-1., -1., -1., -1., -1.], dtype=float32)]
msg = '{} requires ndarray or scalar arguments, got {} at position {}.'

    def check_arraylike(fun_name: str, *args: Any):
      """Check if all args fit JAX's definition of arraylike."""
      assert isinstance(fun_name, str), f"fun_name must be a string. Got {fun_name}"
      if any(not _arraylike(arg) for arg in args):
        pos, arg = next((i, arg) for i, arg in enumerate(args)
                        if not _arraylike(arg))
        msg = "{} requires ndarray or scalar arguments, got {} at position {}."
>       raise TypeError(msg.format(fun_name, type(arg), pos))
E       TypeError: roots requires ndarray or scalar arguments, got <class 'list'> at position 0.
E       Falsifying example: test_jax_roots(
E           on_device='cpu',
E           frontend='jax',
E           backend_fw='numpy',
E           dtype_and_x=(['float32'],
E            [array([-1., -1., -1., -1., -1.], dtype=float32)]),
E           test_flags=FrontendFunctionTestFlags(
E               num_positional_args=0,
E               with_out=False,
E               inplace=False,
E               as_variable=[False],
E               native_arrays=[False],
E               test_compile=False,
E               generate_frontend_arrays=False,
E               transpile=False,
E               precision_mode=False,
E           ),
E           fn_tree='ivy.functional.frontends.jax.numpy.roots',
E       )
E       
E       You can reproduce this example by temporarily adding @reproduce_failure('6.82.3', b'AXicY2AAAUYGBjhNKhsIAAGjAAs=') as a decorator on your test case

/opt/fw/jax/jax/_src/numpy/util.py:328: TypeError

File: ivy_tests/test_ivy/test_frontends/test_jax/test_numpy/test_mathematical_functions.py [hunk 1]
```diff
@@ -2819,6 +2819,40 @@ def test_jax_remainder(
...
+    input_dtype, x = dtype_and_x
+    assume("float16" not in input_dtype)
+    helpers.test_frontend_function(
+        input_dtypes=input_dtype,
+        test_flags=test_flags,
+        frontend=frontend,
+        backend_to_test=backend_fw,
+        fn_tree=fn_tree,
+        on_device=on_device,
+        p=x,
```
Comment (Ookamice) [inline ivy_tests/test_ivy/test_frontends/test_jax/test_numpy/test_mathematical_functions.py:2850]: I believe your issue will be solved with this
```suggestion
        p=x[0],
```

**Comment by dash96** [comment]
How can I work with dtype errors. During all the tests I am getting the right answers but not the right dtypes. When ground truth is `complex64` I am getting `complex128` and when I hard code the `dtype="complex64"` in the function, for the same test cases ground truth is changing to `complex128`

**Comment by dash96** [comment]
Hello! @ZoeCD all the tests are passing in my system. Could you please help me understand what shall I do next.

**Comment by dash96** [comment]
Hello @DecFox Please help me in merging this PR all the tests are passing from my end

**Comment by dash96** [comment]
Hello @hirwa-nshuti could you please help me merge this function it has been here since a month.

File: ivy_tests/test_ivy/test_frontends/test_jax/test_numpy/test_mathematical_functions.py [hunk 2]
```diff
@@ -2819,6 +2819,45 @@ def test_jax_remainder(
...
+            test_flags=test_flags,
+            frontend=frontend,
+            backend_to_test=backend_fw,
+            fn_tree=fn_tree,
+            on_device=on_device,
+            p=x[0],
+            atol=1e-05,
+            rtol=1e-03,
+        )
+        ret, frontend_ret = call()
```
Comment (Ishticode) [inline ivy_tests/test_ivy/test_frontends/test_jax/test_numpy/test_mathematical_functions.py:2855]: unless test_values=False, the test_frontend_function doesn't return the values. In addition this `call()` should be made out side the the definition of `call`. Tbh we don't even need to define it as a call. But thats fine.
Once this is updated pls request my review. Thnx

**Review by Ishticode — CHANGES_REQUESTED** [review]

**Comment by ivy-seed** [comment]
This PR has been labelled as stale because it has been inactive for more than 7 days. If you would like to continue working on this PR, then please add another comment or this PR will be closed in 7 days.

**Comment by ivy-seed** [comment]
This PR has been closed because it has been marked as stale for more than 7 days with no activity.

**Pull request closed without merging**